src/test/scala/uk/gov/nationalarchives/dp/client/DataProcessorTest.scala

```diff
@@ -0,0 +1,216 @@
+    val input =
+      <MetadataResponse>
+        <AdditionalInformation>
+          <Metadata>
+          </Metadata>
+        </AdditionalInformation>
+      </MetadataResponse>
+    val response = new DataProcessor[F]().fragmentUrls(input)
+    val values = valueFromF(response)
+
+    values.size should equal(0)
+  }
+
+  "fragments" should "return multiple fragments" in {
+    def fragment(i: Int) = <Fragment>Fragment
+      {i}
+    </Fragment>
+
+    def fragmentContainer(i: Int) =
+      <MetadataResponse>
+        <MetadataContainer>
+          <Content>
+            {fragment(i)}
+          </Content>
+        </MetadataContainer>
+      </MetadataResponse>
+
+    val fragmentsF =
+      new DataProcessor[F]().fragments(Seq(fragmentContainer(1), fragmentContainer(2)))
+    val fragments = valueFromF(fragmentsF)
+
+    fragments.size should equal(2)
+    fragments.head.trim should equal(fragment(1).toString)
+    fragments.last.trim should equal(fragment(2).toString)
+  }
+
+  "fragments" should "return an empty list if there is no content" in {
+    val input =
+      <MetadataResponse>
+        <MetadataContainer>
+        </MetadataContainer>
+      </MetadataResponse>
+
+    val fragmentsF = new DataProcessor[F]().fragments(Seq(input))
+    val fragments = valueFromF(fragmentsF)
+
+    fragments.size should equal(0)
+  }
+
+  "generationUrlFromEntity" should "return a generation url" in {
+    val input = <EntityResponse>
+      <AdditionalInformation>
+        <Generations>http://localhost:test</Generations>
+      </AdditionalInformation>
+    </EntityResponse>
+    val generationsF = new DataProcessor[F]().generationUrlFromEntity(input)
+    val generations = valueFromF(generationsF)
+
+    generations should equal("http://localhost:test")
+  }
+
+  "generationUrlFromEntity" should "raise an error if the generation is not found" in {
+    val input = <EntityResponse>
+      <AdditionalInformation>
+      </AdditionalInformation>
+    </EntityResponse>
+    val generationsF = new DataProcessor[F]().generationUrlFromEntity(input)
+    val generationsError = intercept[Throwable] {
+      valueFromF(generationsF)
+    }
+    generationsError.getMessage should equal("Generation not found")
+  }
+
+  "allGenerationUrls" should "return a sequence of generation urls" in {
+    val input =
+      <GenerationsResponse>
+        <Generations>
+          <Generation>http://localhost/generation1</Generation>
+          <Generation>http://localhost/generation2</Generation>
+        </Generations>
+      </GenerationsResponse>
+    val generationsF = new DataProcessor[F]().allGenerationUrls(input)
+    val generations = valueFromF(generationsF)
+
+    generations.size should equal(2)
+    generations.head should equal("http://localhost/generation1")
+    generations.last should equal("http://localhost/generation2")
+  }
+
+  "allGenerationUrls" should "return an empty sequence if there are no generation urls" in {
```

Comment:
Is this possible? And if so, should this be an error?

Reply:
I don't think this is possible either. There has to be at least one generation I think? I've updated the code to return an error and changed the test.